src/network/impl/dialer_impl.cpp

```diff
@@ -3,116 +3,142 @@
+          auto ctx_found = dialing_peers_.find(peer_id);
+          if (dialing_peers_.end() == ctx_found) {
+            SL_ERROR(
+                log_,
+                "State inconsistency - uninteresting dial result for peer {}",
+                peer_id.toBase58());
+            if (result.has_value() and not result.value()->isClosed()) {
+              auto close_res = result.value()->close();
+              BOOST_ASSERT(close_res);
+            }
             return;
           }
 
-          if (connection_result) {
-            // we've got the first successful connection to the peer, hooray!
-            ctx->connected = true;
-            // allow the connection accept inbound streams
-            listener->onConnection(connection_result);
-            // return connection to the user
-            ctx->cb(connection_result.value());
+          if (result.has_value()) {
+            listener_->onConnection(result);
+            completeDial(peer_id, result);
             return;
           }
 
-          // here we handle failed attempt to connect
-          if (0 == ctx->calls_remain) {
-            // that was the last attempt to connect and we are still not
-            // connected so lets report an error to the user
-            ctx->cb(connection_result.error());
-            return;
-          }  // otherwise we don't care about this particular failure because at
-             // the end we either would get connected or will do the same -
-             // report the error to the user inside the callback of the last
-             // dial attempt
+          // store an error otherwise and reschedule one more rotate
+          ctx_found->second.result = std::move(result);
+          scheduler_->schedule([this, peer_id] { rotate(peer_id); });
         };
 
-    bool dialled{false};
-    // for all multiaddresses supplied in peerinfo
-    for (auto &&ma : p.addresses) {
-      // try to find best possible transport
-      if (auto tr = this->tmgr_->findBest(ma); tr != nullptr) {
-        // we can dial to this peer!
-        dialled = true;
-        // dial using best transport
-        tr->dial(p.id, ma, dial_handler, timeout);
-        // All the dials are still to be executed sequentially within the single
-        // boost::asio::io_context. Immediate spawn of all of them allows us to
-        // have the closest timeout to the specified instead of
-        // NumberOfMultiaddresses multiplied by a single timeout value.
-      }
+    auto first_addr = ctx.addresses.begin();
+    const auto addr = *first_addr;
+    ctx.tried_addresses.insert(addr);
+    ctx.addresses.erase(first_addr);
+    if (auto tr = tmgr_->findBest(addr); nullptr != tr) {
+      ctx.dialled = true;
+      SL_TRACE(log_, "Dial to {} via {}", peer_id.toBase58().substr(46),
+               addr.getStringAddress());
+      tr->dial(peer_id, addr, dial_handler, ctx.timeout);
+    } else {
+      scheduler_->schedule([this, peer_id] { rotate(peer_id); });
```

Comment:
It can be dangerous to use object is external for lambda asynchronously. Think about using of weak pointers